File: benchmark/comparison/results/haiku_ts/08.ts
type Expr =
  | { type: "Num"; value: number }
  | { type: "Add"; a: Expr; b: Expr }
  | { type: "Mul"; a: Expr; b: Expr }
  | { type: "Neg"; x: Expr };

function evalExpr(expr: Expr): number {
  switch (expr.type) {
    case "Num":
      return expr.value;
    case "Add":
      return evalExpr(expr.a) + evalExpr(expr.b);
    case "Mul":
      return evalExpr(expr.a) * evalExpr(expr.b);
    case "Neg":
      return -evalExpr(expr.x);
  }
}

function show(expr: Expr): string {
  switch (expr.type) {
    case "Num":
      return expr.value.toString();
    case "Add":
      return `(${show(expr.a)} + ${show(expr.b)})`;
    case "Mul":
      return `(${show(expr.a)} * ${show(expr.b)})`;
    case "Neg":
      return `(-(${show(expr.x)}))`;
  }
}

// Test cases
const tests: Expr[] = [
  { type: "Num", value: 5 },
  { type: "Add", a: { type: "Num", value: 3 }, b: { type: "Num", value: 5 } },
  { type: "Mul", a: { type: "Num", value: 2 }, b: { type: "Num", value: 4 } },
  { type: "Neg", x: { type: "Num", value: 3 } },
  { type: "Neg", x: { type: "Add", a: { type: "Num", value: 2 }, b: { type: "Num", value: 3 } } },
];

for (const expr of tests) {
  console.log(evalExpr(expr));
}

for (const expr of tests) {
  console.log(show(expr));
}

File: benchmark/comparison/results/adv_py/08.py
class Expr:
    pass

class Num(Expr):
    def __init__(self, n):
        self.n = n

class Add(Expr):
    def __init__(self, a, b):
        self.a = a
        self.b = b

class Mul(Expr):
    def __init__(self, a, b):
        self.a = a
        self.b = b

class Neg(Expr):
    def __init__(self, x):
        self.x = x

def eval_expr(e):
    if isinstance(e, Num):
        return e.n
    elif isinstance(e, Add):
        return eval_expr(e.a) + eval_expr(e.b)
    elif isinstance(e, Mul):
        return eval_expr(e.a) * eval_expr(e.b)
    elif isinstance(e, Neg):
        return -eval_expr(e.x)

def show(e):
    if isinstance(e, Num):
        return str(e.n)
    elif isinstance(e, Add):
        return f"({show(e.a)} + {show(e.b)})"
    elif isinstance(e, Mul):
        return f"({show(e.a)} * {show(e.b)})"
    elif isinstance(e, Neg):
        return f"(-({show(e.x)}))"

eval_cases = [Num(5), Add(Num(3), Num(5)), Neg(Num(3)), Neg(Add(Num(2), Num(3)))]
for e in eval_cases:
    print(eval_expr(e))

show_cases = [Num(5), Add(Num(3), Num(5)), Mul(Num(2), Num(4)), Neg(Num(3)), Neg(Add(Num(2), Num(3)))]
for e in show_cases:
    print(show(e))
